cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -427,7 +433,51 @@ template <>
 hash_value_type __device__ inline SparkMurmurHash3_32<numeric::decimal128>::operator()(
   numeric::decimal128 const& key) const
 {
-  return this->compute<__int128_t>(key.value());
+  // Generates the Spark MurmurHash3 hash value, mimicking the conversion:
+  // java.math.BigDecimal.valueOf(unscaled_value, _scale).unscaledValue().toByteArray()
+  // https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L381
+  __int128_t const val     = key.value();
+  cudf::size_type key_size = sizeof(__int128_t);
+  std::byte const* data    = reinterpret_cast<std::byte const*>(&val);
+
+  // Extract the first bit of the key, which holds the sign.
+  std::byte const sign_bit = data[key_size - 1] & static_cast<std::byte>(0x80);
+
+  // Small negative values start with 0xff..., small positive values start with 0x00...
+  std::byte const zero_value = std::to_integer<uint8_t>(sign_bit) ? static_cast<std::byte>(0xff)
+                                                                  : static_cast<std::byte>(0x00);
+  // Special cases for 0 and -1 which do not shorten correctly.
+  if (val == 0) { return this->compute(static_cast<uint8_t>(0)); }
+  if (val == static_cast<__int128>(-1)) { return this->compute(static_cast<uint8_t>(0xFF)); }
+
+  // Searching from the big-byte end, find the first non-zero byte in the unscaled little endian
+  // value. The preceding zero bytes that are bigger-bytes in the endian value are not hashed.
```

Comment:
This sentence "The preceding zero bytes that are bigger-bytes in the endian value are not hashed." doesn't make sense to me. Are you missing a "little" before "endian"? Maybe try something like:

> If the value can be represented with a shorter integer type, the leading bytes of the little-endian value are truncated and not included in the hash.